### PR TITLE
Expose `batched_reduce_size` via `_search`

### PR DESCRIPTION
**Pull request by s1monw**
In #23253 we added the ability to incrementally reduce search results.
This change exposes the parameter to control the batch size and therefore
the memory consumption of a large search request.

### CODE REVIEW

File: rest-api-spec/src/main/resources/rest-api-spec/api/search.json [hunk 1]
```diff
@@ -158,6 +158,10 @@
         "request_cache": {
           "type" : "boolean",
           "description" : "Specify if request cache should be used for this request or not, defaults to index level setting"
+        },
+        "batched_reduce_size" : {
+          "type" : "number",
+          "description" : "The number of shard results that should be reduced at once on the coordinating node. This value should be used as a protection mechanism to reduce the memory overhead per search request if the potential number of shards in the request can be large."
```
Comment (clintongormley) [inline rest-api-spec/src/main/resources/rest-api-spec/api/search.json]: add

    "default" : 512

**Review by clintongormley — COMMENTED** [review]
LGTM, but it's missing the doc changes.

**Comment by s1monw** [comment]
> LGTM, but it's missing the doc changes.

I am not sure what you mean? I didn't add a documentation for this since it's so specialized. I want to expose it once we remove the softlimit?

File: test/framework/src/main/java/org/elasticsearch/test/client/RandomizingClient.java [hunk 2]
```diff
@@ -60,7 +60,7 @@ public RandomizingClient(Client client, Random random) {
 
     @Override
     public SearchRequestBuilder prepareSearch(String... indices) {
-        return in.prepareSearch(indices).setSearchType(defaultSearchType).setPreference(defaultPreference).setReduceUpTo(reduceUpTo);
+        return in.prepareSearch(indices).setSearchType(defaultSearchType).setPreference(defaultPreference).setBatchedReduceSize(reduceUpTo);
```
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/client/RandomizingClient.java]: Do you want to rename `reduceUpTo` here as well?

**Review by nik9000 — APPROVED** [review]
I think it'd be useful to return the number of reduction phases that we did so we can add a test that asserts that we did the right number. Just to make sure we didn't drop setting the parameter on the floor. I'm fine with you doing that in a followup or doing it myself since I'm the one that wants it.

**Comment by s1monw** [comment]
> I think it'd be useful to return the number of reduction phases that we did so we can add a test that asserts that we did the right number. Just to make sure we didn't drop setting the parameter on the floor. I'm fine with you doing that in a followup or doing it myself since I'm the one that wants it.

I don't think we should clutter the API with such an internal optimization. What do you expect from it?

**Comment by nik9000** [comment]
> What do you expect from it?

Just to know if the parameter had any effect at all. Right now you can't tell.

**Comment by s1monw** [comment]
> Just to know if the parameter had any effect at all. Right now you can't tell.

this argument is odd. We are testing it throughout the stack with unittests and we know it's passed to the SearchRequest since there is the exception coming from. We can't pass any pointers back just for the integration test sake?

**Comment by nik9000** [comment]
Once we've removed the limit we can test with a really large reduce. We'll want to do that anyway. I like returning the reduction count as well because it is simpler to debug if it fails and because it'll give us more information if the huge reduce fails and the reduction count test doesn't. I'm ok with not doing the test.

**Comment by s1monw** [comment]
> Once we've removed the limit we can test with a really large reduce. We'll want to do that anyway. I like returning the reduction count as well because it is simpler to debug if it fails and because it'll give us more information if the huge reduce fails and the reduction count test doesn't. I'm ok with not doing the test.

we have a whole bunch of tests for this that were added in the original PR

**Comment by s1monw** [comment]
I spoke to @nik9000 and I start to agree we should have this response parameter especially for debugging purposes of this feature at the users end. I added new commits.

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java [hunk 3]
```diff
@@ -179,13 +179,6 @@ public void scrollId(String scrollId) {
         return internalResponse.profile();
     }
 
-    static final class Fields {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: :heart:

**Review by nik9000 — APPROVED** [review]

**Comment by clintongormley** [comment]
> I am not sure what you mean? I didn't add a documentation for this since it's so specialized. I want to expose it once we remove the softlimit?

OK

**Review by clintongormley — APPROVED** [review]

File: core/src/main/java/org/elasticsearch/action/search/SearchResponse.java [hunk 4]
```diff
@@ -189,14 +190,18 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
...
-        builder.field(Fields.TOOK, tookInMillis);
-        builder.field(Fields.TIMED_OUT, isTimedOut());
+        builder.field("took", tookInMillis);
+        builder.field("timed_out", isTimedOut());
         if (isTerminatedEarly() != null) {
-            builder.field(Fields.TERMINATED_EARLY, isTerminatedEarly());
+            builder.field("terminated_early", isTerminatedEarly());
+        }
+        if (getNumReducePhases() != 1) {
+            builder.field("num_reduce_phases", getNumReducePhases());
```
Comment (Mpdreamz) [inline core/src/main/java/org/elasticsearch/action/search/SearchResponse.java]: :heart:

**Pull request merged**